### DISCUSSION TITLE
Add support for bun

### DISCUSSION THREAD

**Discussion by SimonsThijs** [medusajs/medusa#5036] [Feature Requests]
Bun has recently launched with the goal to create a more user friendly js/ts development experience. It could also be a lot faster and more memory efficient, e.g., i sometimes run out of memory on my 8gb m1 macbook pro when i am running the backend and frontend both in development mode at the same time. 

Installations of packages should be a lot faster as well with bun which is especially nice for deployments.

In general bun should give a much cleaner and less complex development experience and i think medusa could profit from this.

Has anyone already tried running medusa with bun?

**Comment by srindom** [comment]
I used bun to set up a project and install dependencies with `bun create medusa-app`. Required no additional changes.

Haven't looked into what would be required to swap node with bun though.

**Reply by SimonsThijs** [reply]
Some parts of medusa might still rely on node api's that have not yet been implemented. See this page for node compatibility: https://bun.sh/docs/runtime/nodejs-apis#node-fs

For example, next.js can not yet be run with the bun runtime and still relies on node: https://bun.sh/guides/ecosystem/nextjs

**Reply by pevey** [reply]
If you join the Medusa Discord, there is already some discussion of this.  The Medusa backend appears to already be fully compatible.  The frontend is totally up to the user.

https://discord.com/channels/876835651130097704/899713621129564200/1150160296594374817

**Reply by helpisdev** [reply]
Just setup a fresh server instance to experiment with bun/medusa/htmx/percona. If npm isn't installed, it does not work out of the box with bun. Particularly it fails quite early with:

```sh
An error occurred while preparing project: Error: Command failed: npm install --legacy-peer-deps
```

**Comment by BleedingDev** [comment]
What's the status of this? Bun is now way more stable and way more compatible with Node.js ecosystem, yet there are some questions. We did some analysis and so far (AI) says:

<details>
<summary>Details</summary>

● Based on the exploration, here's what would be required to switch from Node to Bun:

  Key Node-specific Dependencies

  The develop command relies heavily on these Node APIs:

  1. child_process.fork() - Spawns the dev server as a child process with IPC
  2. cluster module - Multi-process worker management
  3. require.cache manipulation - For hot module reloading
  4. Process IPC - process.send() / process.on("message") for HMR communication

  What Would Need to Change for Bun

  1. Child Process Spawning

  Bun supports child_process.fork() but with some differences. You'd need to test IPC communication compatibility.

  2. Cluster Module

  Bun does not support the cluster module. You'd need to:
  - Remove cluster support entirely, or
  - Reimplement using Bun's native worker threads (Bun.spawn with workers)

  3. Module Cache (HMR)

  Bun uses a different module system. Instead of require.cache, you'd use:
  - Bun.Transpiler for cache invalidation
  - Or leverage Bun's built-in --hot flag which has native HMR

  4. File Watching

  Bun has native Bun.FileSystemWatcher which could replace chokidar.

  Practical Path Forward

  The simplest approach:

  # Bun can run Node apps with compatibility mode
  bun --bun packages/medusa/src/commands/develop.ts

  However, for full Bun optimization, you'd need to modify:
  - packages/medusa/src/commands/develop.ts - Replace fork() with Bun.spawn()
  - packages/medusa/src/commands/start.ts - Remove cluster usage or use Bun workers
  - packages/medusa/src/commands/utils/dev-server/module-cache-manager.ts - Adapt to Bun's module system

Production Start - Node Dependencies

  The start command (packages/medusa/src/commands/start.ts) still uses:

  1. cluster module - When --cluster flag is used for multi-process scaling
  2. Standard HTTP/Express - Bun-compatible
  3. os.cpus() - Bun-compatible
  4. Signal handlers - Bun-compatible

  Running Production with Bun

  Without clustering - Should work:
  cd .medusa/server
  bun run node_modules/@medusajs/medusa/dist/commands/start.js
  # or if medusa CLI is available
  bun medusa start

  With clustering (--cluster flag) - Won't work due to cluster module.

  Quick Test

  You could try it now on an existing build:

  # Build normally
  yarn medusa build

  # Try running with Bun
  cd .medusa/server
  bun --bun node_modules/.bin/medusa start

  The main unknowns would be:
  - Database drivers (pg/postgres) compatibility with Bun
  - Any native Node addons in dependencies
  - MikroORM/TypeORM behavior under Bun

  If you're not using --cluster, there's a good chance it works out of the box. Want me to check if there are any other Node-specific APIs in the core framework loaders that might cause issues?

</details>

**Comment by picardplaisimond** [comment]
I'm running medusa with bun and so far it's working well. The only thing is that you have to setup your project manually.

**Reply by valiukasd** [reply]
Are you using the Bun runtime? Last time I checked, Bun can install packages if set up with correct package hoisting settings, but I could not get the dev server started with the --bun flag. Have you got it working?

**Reply by picardplaisimond** [reply]
Hmmm... with the version 2.13.1, it's not working anymore with Bun...

**Edit:**
Sorry, I was talking about about Bun as a package manager.

**Comment by picardplaisimond** [comment]
Since the version 2.13.0, MedusaJS is not working well with Bun anymore (in a monorepo). The breaking change with zod did something...
I'm not sure if the problem is from Medusa or from Bun.

Even with the isolated mode : `bun install --linker isolated`, the problem is still the same.
My whole monorepo is working with bun... but I may have to switch back to pnpm :(

```
error:   Error starting server
error:   An error occurred while registering API Routes. Error: Cannot read properties of undefined (reading 'def')
```

---------------------------
**Edit:**
The error is still happening with pnpm. But the error is not happening in a normal repo, only in monorepo with other apps.
More logs on that errors:

```
Error: An error occurred while registering API Routes. Error: Cannot read properties of undefined (reading 'def')
at exports.default (monorepo\node_modules\.bun\@medusajs+medusa@2.13.1+0fde9bf9044c4459\node_modules\@medusajs\medusa\src\loaders\api.ts:72:11)
at async Promise.allSettled (index 2)
at async promiseAll (monorepo\node_modules\.bun\@medusajs+utils@2.13.1+c946557b397becc3\node_modules\@medusajs\utils\src\common\promise-all.ts:27:18)
at async loadEntrypoints (monorepo\node_modules\.bun\@medusajs+medusa@2.13.1+0fde9bf9044c4459\node_modules\@medusajs\medusa\src\loaders\index.ts:117:26)
at async exports.default (monorepo\node_modules\.bun\@medusajs+medusa@2.13.1+0fde9bf9044c4459\node_modules\@medusajs\medusa\src\loaders\index.ts:219:7)
at async internalStart (monorepo\node_modules\.bun\@medusajs+medusa@2.13.1+0fde9bf9044c4459\node_modules\@medusajs\medusa\src\commands\start.ts:273:59)
at async start (monorepo\node_modules\.bun\@medusajs+medusa@2.13.1+0fde9bf9044c4459\node_modules\@medusajs\medusa\src\commands\start.ts:419:12)
```

**Reply by BleedingDev** [reply]
I think this ticket should be separated into bun runtime and bun package manager.
